Initial porous flow 6845 - poroperm functions

hsheldon opened the pull request:
Permeability as a function of porosity with some simple tests. Please ignore stupid commit message "Porosity as a function of permeability", obviously I meant "permeability as a function of porosity"!

hsheldon committed:
Porosity as a function of permeability. Still need to do _dpermeability_dvar.

hsheldon committed:
Working on the derivatives, need to multiply by dporosity_dvar.

hsheldon committed:
Completed permeability from porosity, including trivial tests

hsheldon committed:
Correcting error with quote marks

WilkAndy commented:
Sorry, heather, i'm taking today off, so haven't replied to your other questions.  In about an hour i'll have a look over your code.  Did you mean to put a PR to my personal moose repository (as you have done) or into idaholab/moose?

a

hsheldon commented:
Um, good question, I was working off your porous flow 6845 branch hence I pushed to your repository but I suppose I should be forking from idaholab/moose now. I’ll do that.

From: Andy Wilkins [mailto:notifications@github.com]
Sent: Monday, 1 August 2016 12:34 PM
To: WilkAndy/moose moose@noreply.github.com
Cc: Sheldon, Heather (Mineral Resources, Kensington WA) Heather.Sheldon@csiro.au; Author author@noreply.github.com
Subject: Re: [WilkAndy/moose] Initial porous flow 6845 - poroperm functions (#45)

Sorry, heather, i'm taking today off, so haven't replied to your other questions. In about an hour i'll have a look over your code. Did you mean to put a PR to my personal moose repository (as you have done) or into idaholab/moose?

a

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHubhttps://github.com/WilkAndy/moose/pull/45#issuecomment-236487276, or mute the threadhttps://github.com/notifications/unsubscribe-auth/AKJ_8nHWV9C_M1YfJkU2EMpLqH3GONyTks5qbXdGgaJpZM4JZPBg.

hsheldon commented:
Ignore this - I get it now - its moved into porosityUnity.

Now I’m confused. There are some differences between idaholab/moose and your repository branch porous flow 6845, which don’t make sense and I can’t see how or when these differences came in.

A particular problem is PorousFlowPorosityHM.C. It is quite different in the two repositories. For example, in your respository it has declareProperty for _porosity_nodal, _porosity_qp etc. But in idaholab/moose it does not have the declareProperty lines. There are some other differences too.

What’s going on?

From: Andy Wilkins [mailto:notifications@github.com]
Sent: Monday, 1 August 2016 12:34 PM
To: WilkAndy/moose moose@noreply.github.com
Cc: Sheldon, Heather (Mineral Resources, Kensington WA) Heather.Sheldon@csiro.au; Author author@noreply.github.com
Subject: Re: [WilkAndy/moose] Initial porous flow 6845 - poroperm functions (#45)

Sorry, heather, i'm taking today off, so haven't replied to your other questions. In about an hour i'll have a look over your code. Did you mean to put a PR to my personal moose repository (as you have done) or into idaholab/moose?

a

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHubhttps://github.com/WilkAndy/moose/pull/45#issuecomment-236487276, or mute the threadhttps://github.com/notifications/unsubscribe-auth/AKJ_8nHWV9C_M1YfJkU2EMpLqH3GONyTks5qbXdGgaJpZM4JZPBg.

WilkAndy commented on modules/porous_flow/include/materials/PorousFlowMaterialPermeabilityFromPorosity.h:
```diff
@@ -0,0 +1,76 @@
+
+/**
+ * Material designed to provide the permeability tensor
+ * which is a function of porosity
```
Here you should explain the functional permeability = porosity^n/(1 - porosity)^m, etc

WilkAndy commented on modules/porous_flow/include/materials/PorousFlowMaterialPermeabilityFromPorosity.h:
```diff
@@ -0,0 +1,76 @@
+#include "PorousFlowDictator.h"
+
+//Forward Declarations
+class PorousFlowMaterialPermeabilityFromPorosity;
```
I suggest calling this PorousFlowPermeabilityKozenyCarman .  Other people might want other permeability-porosity relationships

WilkAndy commented on modules/porous_flow/include/materials/PorousFlowMaterialPermeabilityFromPorosity.h:
```diff
@@ -0,0 +1,76 @@
+ * Material designed to provide the permeability tensor
+ * which is a function of porosity
+ */
+class PorousFlowMaterialPermeabilityFromPorosity : public DerivativeMaterialInterface<Material>
```
I suggest deriving from PorousFlowPermeabilityConst (or creating a new PorousFlowPermeabilityUnity and subclassing the "Const" and "KozenyCarman" from it).  That will be more ordered.

WilkAndy commented on modules/porous_flow/src/materials/PorousFlowMaterialPermeabilityFromPorosity.C:
```diff
@@ -0,0 +1,82 @@
+  params.addRequiredParam<Real>("n", "Porosity exponent (numerator)");
+  params.addRequiredParam<Real>("m", "(1-porosity) exponent (denominator)");
+  params.addRequiredParam<UserObjectName>("PorousFlowDictator_UO", "The UserObject that holds the list of Porous-Flow variable names.");
+  params.addClassDescription("This Material calculates the permeability tensor from a specified function of porosity");
```
Specify the actual functions used here?

WilkAndy commented on modules/porous_flow/src/materials/PorousFlowMaterialPermeabilityFromPorosity.C:
```diff
@@ -0,0 +1,82 @@
+}
+
+void
+PorousFlowMaterialPermeabilityFromPorosity::initQpStatefulProperties()
```
It's probably not necessary to have initQpStatefulProperties - it is for properties (such as porosity) that need "old" values.  Permeability doesn't, so you probably don't need initQp...  However, having said this, there is some bug that's yet to be resolved in MOOSE regarding this stuff, so try to remove initQp and if you get a crash (in debug mode) just put it back in.

WilkAndy commented on modules/porous_flow/src/materials/PorousFlowMaterialPermeabilityFromPorosity.C:
```diff
@@ -0,0 +1,82 @@
+
+  _dpermeability_dvar[_qp].resize(_num_var, RealTensorValue());
+  for (unsigned v = 0; v < _num_var; ++v)
+    _dpermeability_dvar[_qp][v] = _dporosity_qp_dvar[_qp][v]*_permeability[_qp] * (_n/_porosity_qp[_qp] + _m/(1.0-_porosity_qp[_qp]));
```
This derivative looks correct.  Please could you check the MOOSE Jacobian using the snes_type=test using the PorousFlowAdvectiveFlux.  Look at the tests/jacobian/fflux*.i to copy.  The reason i'm asking is that Kernel contains dpermeability_dvar, but until now it has always been zero, so the Kernel Jacobian itself might actually be wrong, even if your derivative is correct (or the Kernel Jacobian might be right and your derivative wrong, or both wrong, or both right!)

hsheldon replied:
Would the test need to have porosity changing with time, otherwise there would be no point?

From: Andy Wilkins [mailto:notifications@github.com]
Sent: Monday, 1 August 2016 1:45 PM
To: WilkAndy/moose moose@noreply.github.com
Cc: Sheldon, Heather (Mineral Resources, Kensington WA) Heather.Sheldon@csiro.au; Author author@noreply.github.com
Subject: Re: [WilkAndy/moose] Initial porous flow 6845 - poroperm functions (#45)

In modules/porous_flow/src/materials/PorousFlowMaterialPermeabilityFromPorosity.Chttps://github.com/WilkAndy/moose/pull/45#discussion_r72926260:

> +
> 
> +void
> 
> +PorousFlowMaterialPermeabilityFromPorosity::initQpStatefulProperties()
> 
> +{
> -  // *********\*  Not sure if this is the right thing or even possible! If not need to figure out an alternative for initialising perm.
> -  _permeability[_qp] = _k_anisotropy__mult_pow(_porosity_qp[_qp],_n)/pow(1.0-_porosity_qp[_qp],_m);
> 
> +}
> 
> +
> 
> +void
> 
> +PorousFlowMaterialPermeabilityFromPorosity::computeQpProperties()
> 
> +{
> -  _permeability[_qp] = _k_anisotropy__mult_pow(_porosity_qp[_qp],_n)/pow(1.0-_porosity_qp[_qp],_m);
> 
> +
> -  _dpermeability_dvar[_qp].resize(_num_var, RealTensorValue());
> -  for (unsigned v = 0; v < _num_var; ++v)
> -    _dpermeability_dvar[_qp][v] = _dporosity_qp_dvar[_qp][v]*_permeability[_qp] \* (_n/_porosity_qp[_qp] + _m/(1.0-_porosity_qp[_qp]));

This derivative looks correct. Please could you check the MOOSE Jacobian using the snes_type=test using the PorousFlowAdvectiveFlux. Look at the tests/jacobian/fflux*.i to copy. The reason i'm asking is that Kernel contains dpermeability_dvar, but until now it has always been zero, so the Kernel Jacobian itself might actually be wrong, even if your derivative is correct (or the Kernel Jacobian might be right and your derivative wrong, or both wrong, or both right!)

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHubhttps://github.com/WilkAndy/moose/pull/45/files/70988634d2f96dd7869d443ace65039c2244f23d#r72926260, or mute the threadhttps://github.com/notifications/unsubscribe-auth/AKJ_8hGIJziUgP-uplqGHJpI5aSur8Reks5qbYfNgaJpZM4JZPBg.

WilkAndy replied:
Maybe i should do the Jacobian test.  It doens't need to be a change-in-time thing.  What PETSc does is vary the nodal displacements, which will vary the porosity, which will vary the permeability, which will vary the residual, which will go into the jacobian.  but there's tonnes of places where it can all go wrong, and i don't really want you to bang you head against the desk for a few days getting it right, or finally understanding that it was all correct initially but you didn't realise.

WilkAndy commented on modules/porous_flow/src/materials/PorousFlowMaterialPermeabilityFromPorosity.C:
```diff
@@ -0,0 +1,82 @@
+PorousFlowMaterialPermeabilityFromPorosity::initQpStatefulProperties()
+{
+  // **********  Not sure if this is the right thing or even possible! If not need to figure out an alternative for initialising perm.
+  _permeability[_qp] = _k_anisotropy*_mult*pow(_porosity_qp[_qp],_n)/pow(1.0-_porosity_qp[_qp],_m);
```
Operator spacing.  MOOSE has adopted "a \* b + c / d" instead of "a*b+c/d" - ie, a space around all arithmetic operators.

WilkAndy commented on modules/porous_flow/src/materials/PorousFlowMaterialPermeabilityFromPorosity.C:
```diff
@@ -0,0 +1,82 @@
+    _d( _d_set ? getParam<Real>("d") : -1 ),
+    _m(getParam<Real>("m")),
+    _n(getParam<Real>("n")),
+    _k_anisotropy( _k_anisotropy_set ? getParam<RealTensorValue>("k_anisotropy") : getParam<RealTensorValue>("1 0 0  0 1 0  0 0 1")),
```
Hopefully you can derive from PorousFlowPermeabilityConstant (or Unity, as described above) and you won't need this parameter

WilkAndy commented on modules/porous_flow/src/materials/PorousFlowMaterialPermeabilityFromPorosity.C:
```diff
@@ -0,0 +1,82 @@
+
+PorousFlowMaterialPermeabilityFromPorosity::PorousFlowMaterialPermeabilityFromPorosity(const InputParameters & parameters) :
+    DerivativeMaterialInterface<Material>(parameters),
+    _k0_set( parameters.isParamValid("k0") ),
```
Spacing here is different than below.  I suggest using _k0_set(parameters.isParamValid("k0"))  with no spaces - seems to be MOOSE convention.  Sorry for being so pedantic, but when you put this all into MOOSE you'll get the same comment!

WilkAndy commented on modules/porous_flow/include/materials/PorousFlowMaterialPermeabilityFromPorosity.h:
```diff
@@ -0,0 +1,76 @@
+
+protected:
+  /// flags to indicate whether parameters have been set
+  bool _k0_set;
```
you might not need this "*_set" parameters.  You could simply use parameters.isParamValid(...) in your "if" statements in your constructor.  If you still want to retain them, make them

const bool _k0_set;

The "const" indicates that you will never change them after construction, and the compiler will throw an error if you attempt it.

WilkAndy commented on modules/porous_flow/include/materials/PorousFlowMaterialPermeabilityFromPorosity.h:
```diff
@@ -0,0 +1,76 @@
+  
+  /// parameters for poroperm equation
+  const Real _k0;
+  const Real _phi0;
```
You probably want to comment these parameters.  The comment can be the same.  Eg

/// Permeability = porosity^_n / (1 - porosity)^_m , etc

They are for doxygen, so that someone will click on "_phi0" and hope to get an explanation of what it is.

WilkAndy commented on modules/porous_flow/src/materials/PorousFlowMaterialPermeabilityFromPorosity.C:
```diff
@@ -0,0 +1,82 @@
+void
+PorousFlowMaterialPermeabilityFromPorosity::computeQpProperties()
+{
+  _permeability[_qp] = _k_anisotropy*_mult*pow(_porosity_qp[_qp],_n)/pow(1.0-_porosity_qp[_qp],_m);
```
Use std::pow instead of just pow

hsheldon committed:
Starting work on Andy's comments on PR

WilkAndy commented on modules/porous_flow/include/materials/PorousFlowMaterialPermeabilityKozenyCarman.h:
```diff
@@ -0,0 +1,92 @@
+#include "PorousFlowDictator.h"
+
+//Forward Declarations
+class PorousFlowMaterialPermeabilityKozenyCarman;
```
Just call it PorousFlowPermeabilityKozenyCarman (ie, remove the "Material").  We initially had PorousFlowMaterialXXXX everywhere but then decided it was just too long and we remove the "Material".  Perhaps there are still some classes with "Material" in them, ooops, but don't create a new one.

WilkAndy commented on modules/porous_flow/include/materials/PorousFlowMaterialPermeabilityKozenyCarman.h:
```diff
@@ -0,0 +1,92 @@
+  MaterialProperty<std::vector<RealTensorValue> > & _dpermeability_dvar;
+
+  /// Multiplying factor in k = k_ijk * A * phi^n / (1 - phi)^m
+  Real _A;
```
can be const Real _A;

WilkAndy commented on modules/porous_flow/include/materials/PorousFlowMaterialPermeabilityKozenyCarman.h:
```diff
@@ -0,0 +1,92 @@
+ * A = f * d^2
+ * where f is a scalar constant and d is grain diameter.
+ */
+class PorousFlowMaterialPermeabilityKozenyCarman : public DerivativeMaterialInterface<Material>
```
If you derive from PorousFlowPermeabilityConst you won't need _k_anisotropy, _permeability, or _dpermeability_dvar.

WilkAndy commented on modules/porous_flow/src/materials/PorousFlowMaterialPermeabilityKozenyCarman.C:
```diff
@@ -0,0 +1,70 @@
+void
+PorousFlowMaterialPermeabilityKozenyCarman::computeQpProperties()
+{
+  _permeability[_qp] = _k_anisotropy * _A * std::pow(_porosity_qp[_qp],_n)/std::pow(1.0 - _porosity_qp[_qp],_m);
```
have a space after a comma.

WilkAndy commented on modules/porous_flow/src/materials/PorousFlowMaterialPermeabilityUnity.C:
```diff
@@ -0,0 +1,32 @@
+/*             See LICENSE for full restrictions                */
+/****************************************************************/
+
+#include "PorousFlowMaterialPermeabilityUnity.h"
```
Oh, good.  Thanks for doing this "Unity" version.  I didn't see it before i made the comments above.  Remove "Material" from the name.  Also, i don't think it should be registered in PorousFlowApp as i think it's very unlikely anyone will use it in real simulations.  But, if you disagree just register it, no worries.

WilkAndy replied:
You forgot to "git add" the "Unity.h" file

WilkAndy commented on modules/porous_flow/src/materials/PorousFlowMaterialPermeabilityUnity.C:
```diff
@@ -0,0 +1,32 @@
+PorousFlowMaterialPermeabilityUnity::computeQpProperties()
+{
+  _dpermeability_dvar[_qp].resize(_num_var, RealTensorValue());
+  _permeability_qp[_qp] = getParam<RealTensorValue>("1 0 0  0 1 0  0 0 1");
```
Rather than using getParam, use the constructor:

 = RankTwoTensor(RankTwoTensor::initIdentity)

or

 = RankTwoTensor(1, 0, 0, 0, 1, 0, 0, 0, 1)

whichever looks best to you

hsheldon committed:
Further work on Andy's comments on PR. This compiles and runs the Darcy tests, but some other tests (using PermeabilityConst) are broken. Not sure why.

WilkAndy commented on modules/porous_flow/include/materials/PorousFlowPermeabilityConst.h:
```diff
@@ -0,0 +1,39 @@
+#define POROUSFLOWPERMEABILITYCONST_H
+
+#include "DerivativeMaterialInterface.h"
+#include "Material.h"
```
The Material.h, PorousFlowDictator.h and Deriv...h includes are probaly unnecessary as they'll come with PorousFlowPermeabilityUnity.h

WilkAndy commented on modules/porous_flow/include/materials/PorousFlowPermeabilityKozenyCarman.h:
```diff
@@ -0,0 +1,79 @@
+#include "Material.h"
+#include "PorousFlowPermeabilityUnity.h"
+
+#include "PorousFlowDictator.h"
```
Same "include" comment as for the "PermeabilityConst" class.

WilkAndy commented on modules/porous_flow/include/materials/PorousFlowPermeabilityKozenyCarman.h:
```diff
@@ -0,0 +1,79 @@
+  const MooseEnum _poroperm_function;
+
+  /// Multiplying factor in k = k_ijk * A * phi^n / (1 - phi)^m
+  Real _A;
```
I think you can (should) make this const Real _A, right?

hsheldon replied:
_A is currently initialised in the body of the constructor so I can’t make it const. I would have to put the switch statement into the initialisation list. Is that what I should do?

From: Andy Wilkins [mailto:notifications@github.com]
Sent: Friday, 5 August 2016 8:11 AM
To: WilkAndy/moose moose@noreply.github.com
Cc: Sheldon, Heather (Mineral Resources, Kensington WA) Heather.Sheldon@csiro.au; Author author@noreply.github.com
Subject: Re: [WilkAndy/moose] Initial porous flow 6845 - poroperm functions (#45)

In modules/porous_flow/include/materials/PorousFlowPermeabilityKozenyCarman.hhttps://github.com/WilkAndy/moose/pull/45#discussion_r73624453:

> -  const Real _n;
> 
> +
> -  /// Tensor multiplier k_ijk in k = k_ijk \* A \* phi^n / (1 - phi)^m
> -  const RealTensorValue _k_anisotropy;
> 
> +
> -  /// quadpoint porosity
> -  const MaterialProperty<Real> & _porosity_qp;
> 
> +
> -  /// d(quadpoint porosity)/d(PorousFlow variable)
> -  const MaterialPropertystd::vector<Real > & _dporosity_qp_dvar;
> 
> +
> -  /// Name of porosity-permeability relationship
> -  const MooseEnum _poroperm_function;
> 
> +
> -  /// Multiplying factor in k = k_ijk \* A \* phi^n / (1 - phi)^m
> -  Real _A;

I think you can (should) make this const Real _A, right?

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHubhttps://github.com/WilkAndy/moose/pull/45/files/ce366d17c08b92cdb7d58f1273857ae5c8851b49#r73624453, or mute the threadhttps://github.com/notifications/unsubscribe-auth/AKJ_8mz6EM9KyCip1pkoEfZJV0dzbjKhks5qcn97gaJpZM4JZPBg.

WilkAndy replied:
Oh yea, that's right.  I think the switch would look ugly and confusing in the init list, so let's just leave _A not const.

WilkAndy commented on modules/porous_flow/src/base/PorousFlowApp.C:
```diff
@@ -116,7 +118,8 @@ PorousFlowApp::registerObjects(Factory & factory)
   registerMaterial(PorousFlowPorosityHM);
   registerMaterial(PorousFlowMaterialMassFractionBuilder);
-  registerMaterial(PorousFlowMaterialPermeabilityConst);
+  registerMaterial(PorousFlowPermeabilityConst);
```
Since you include PorousFlowPermeabilityUnity, you probably want to registerMaterial(...Unity) too

WilkAndy commented on modules/porous_flow/src/materials/PorousFlowPermeabilityUnity.C:
```diff
@@ -0,0 +1,32 @@
+    DerivativeMaterialInterface<Material>(parameters),
+    _PorousFlow_name_UO(getUserObject<PorousFlowDictator>("PorousFlowDictator_UO")),
+    _num_var(_PorousFlow_name_UO.numVariables()),
+    _permeability_qp(declareProperty<RealTensorValue>("PorousFlow_permeability_qp")),
```
Thanks for calling this "PorousFlow_permeability_qp" - better to have the "_qp" to indicate it's at the quadpoints

WilkAndy commented on modules/porous_flow/src/materials/PorousFlowPermeabilityConst.C:
```diff
@@ -0,0 +1,32 @@
+InputParameters validParams<PorousFlowPermeabilityConst>()
+{
+  InputParameters params = validParams<Material>();
+
```
Extra line not needed.  Also, you have to have validParams<PorousFlowPermeabilityUnity>();

hsheldon replied:
I don’t understand what you mean. I’ve got validParams?

From: Andy Wilkins [mailto:notifications@github.com]
Sent: Friday, 5 August 2016 8:18 AM
To: WilkAndy/moose moose@noreply.github.com
Cc: Sheldon, Heather (Mineral Resources, Kensington WA) Heather.Sheldon@csiro.au; Author author@noreply.github.com
Subject: Re: [WilkAndy/moose] Initial porous flow 6845 - poroperm functions (#45)

In modules/porous_flow/src/materials/PorousFlowPermeabilityConst.Chttps://github.com/WilkAndy/moose/pull/45#discussion_r73625043:

> @@ -0,0 +1,32 @@
> 
> +/****************************************************************/
> 
> +/\* MOOSE - Multiphysics Object Oriented Simulation Environment  */
> 
> +/\*                                                              */
> 
> +/\*          All contents are licensed under LGPL V2.1           */
> 
> +/\*             See LICENSE for full restrictions                */
> 
> +/****************************************************************/
> 
> +
> 
> +#include "PorousFlowPermeabilityConst.h"
> 
> +
> 
> +template<>
> 
> +InputParameters validParams<PorousFlowPermeabilityConst>()
> 
> +{
> -  InputParameters params = validParams<Material>();
> 
> +

Extra line not needed. Also, you have to have validParams();

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHubhttps://github.com/WilkAndy/moose/pull/45/files/ce366d17c08b92cdb7d58f1273857ae5c8851b49#r73625043, or mute the threadhttps://github.com/notifications/unsubscribe-auth/AKJ_8m8fNa3S5_lLrvsCrN8zDU-XyVIJks5qcoEcgaJpZM4JZPBg.

WilkAndy replied:
Sorry, that comment went crazy.  The line should be:

InputParameters params = validParams<PorousFlowPermeaiblityUnity>()

Ie, get the params from the "Unity" version, and then add your extra ones on top of them

WilkAndy commented on modules/porous_flow/src/materials/PorousFlowPermeabilityKozenyCarman.C:
```diff
@@ -0,0 +1,66 @@
+InputParameters validParams<PorousFlowPermeabilityKozenyCarman>()
+{
+  InputParameters params = validParams<Material>();
+
```
extra line not needed, and validParams<PorousFlowPermeabilityUnity>();

WilkAndy commented on modules/porous_flow/src/materials/PorousFlowPermeabilityKozenyCarman.C:
```diff
@@ -0,0 +1,66 @@
+  params.addParam<MooseEnum>("poroperm_function", poroperm_function, "Function relating porosity and permeability. The options are: kozeny_carman_fd2 = f d^2 phi^n/(1-phi)^m (where phi is porosity, f is a scalar constant with typical values 0.01-0.001, and d is grain size). kozeny_carman_phi0 = k0 (1-phi0)^m/phi0^n * phi^n/(1-phi)^m (where phi is porosity, and k0 is the permeability at porosity p ...
+  params.addParam<Real>("k0", "The permeability scalar value (usually in m^2) at the reference porosity, required for kozeny_carman_phi0");
+  params.addParam<RealTensorValue>("k_anisotropy", "A tensor to multiply the calculated scalar permeability, in order to obtain anisotropy if required. Defaults to isotropic permeability if not specified.");
+  params.addParam<Real>("phi0", "The reference porosity, required for kozeny_carman_phi0");
```
Do you want to do a RangeCheck here (and elsewhere)?

params.addRangeCheckedParam<Real>("phi0", "phi0 > 0 & phi0 < 1", "doco")

(note only one "&" sign)

WilkAndy commented on modules/porous_flow/src/materials/PorousFlowPermeabilityUnity.C:
```diff
@@ -0,0 +1,32 @@
+InputParameters validParams<PorousFlowPermeabilityUnity>()
+{
+  InputParameters params = validParams<Material>();
+  params.addClassDescription("This Material calculates the permeability tensor assuming it is equal to 1 0 0  0 1 0  0 0 1");
```
Need to add the PorousFlowDictator_UO param.

WilkAndy commented on modules/porous_flow/src/materials/PorousFlowPermeabilityUnity.C:
```diff
@@ -0,0 +1,32 @@
+PorousFlowPermeabilityUnity::computeQpProperties()
+{
+  _dpermeability_qp_dvar[_qp].resize(_num_var, RealTensorValue());
+  _permeability_qp[_qp] = getParam<RealTensorValue>("1 0 0  0 1 0  0 0 1");
```
I don't like this.  try

 = RealTensorValue(1, 0, 0, 0, 1, 0, 0, 0, 1);

WilkAndy commented on modules/porous_flow/tests/buckley_leverett/bl01.i:
```diff
@@ -114,7 +114,7 @@
   [../]
   [./permeability]
-    type = PorousFlowMaterialPermeabilityConst
+    type = PorousFlowPermeabilityConst
```
Great, when you've completely finished, try running the tests in debug mode

./run_tests -j 4 --dbg

and

./run_tests -j 4 --dbg --heavy

and that's all!  Then you can put in a PR to idaholab/moose (i can help if i'm at work).

hsheldon committed:
Some further corrections:
- Removing unnecessary #include
- Removed PorousFlowPermeabilityUnity from PorousFlowApp.C
- Correcting validParams
- Range checking
- Adding PorousFlowDictator_U0 as required param

hsheldon committed:
Changed initialisation of permeability tensor

WilkAndy commented:
Looks ready to go, i reckon.  Just check that all the tests run.

hsheldon commented:
Ok. I accept your offer to write the Jacobian test! I’m moving it all into a new branch because your initial_porous_flow_6845 is out of date with idaholab. I’ve made an issue number (#7503). I just need to iron out any wrinkles that appear in the new branch.

From: Andy Wilkins [mailto:notifications@github.com]
Sent: Friday, 5 August 2016 11:31 AM
To: WilkAndy/moose moose@noreply.github.com
Cc: Sheldon, Heather (Mineral Resources, Kensington WA) Heather.Sheldon@csiro.au; Author author@noreply.github.com
Subject: Re: [WilkAndy/moose] Initial porous flow 6845 - poroperm functions (#45)

Looks ready to go, i reckon. Just check that all the tests run.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHubhttps://github.com/WilkAndy/moose/pull/45#issuecomment-237746880, or mute the threadhttps://github.com/notifications/unsubscribe-auth/AKJ_8kQ5HwwF9Cj8Z18TO6y4dqajQq2hks5qcq5wgaJpZM4JZPBg.

WilkAndy commented:
This permeability-as-a-function-of-porosity has all been sorted a few months ago.  Closing @hsheldon 's PR to my devel branch.
